### PR TITLE
[SPARK-37623][SPARK-39230][SQL][FOLLOW-UP] Make regr_slope and regr_intercept safe with ANSI mode

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR proposes to make `regr_slope` and `regr_intercept` ANSI-safe by checking zero to avoid divide-by-zero exceptions when ANSI mode is on.

### Why are the changes needed?

To make both expressions working regardless of ANSI mode.

### Does this PR introduce _any_ user-facing change?

No, both functions are not released yet.

### How was this patch tested?

`JDBCV2Suite` already covers it when ASNI mode is on. This was found by a broken test case, see also https://github.com/apache/spark/pull/36773/files#r916607535

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/linearRegression.scala [hunk 1]
```diff
@@ -278,7 +278,8 @@ case class RegrSlope(left: Expression, right: Expression) extends DeclarativeAgg
     covarPop.mergeExpressions ++ varPop.mergeExpressions
 
   override lazy val evaluateExpression: Expression = {
-    If(covarPop.n === 0.0, Literal.create(null, DoubleType), covarPop.ck / varPop.m2)
+    If(Or(covarPop.n === 0.0, varPop.m2 === 0.0),
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/linearRegression.scala:281]: This is a behavior change in ANSI mode but looks reasonable. The two functions are new in 3.4 so no breaking change. @beliefer can you double-check the formular? I'm wondering why we only checked `covarPop.n === 0.0` at the first place.

Comment (srowen) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/linearRegression.scala:281]: I think we need to remove the check on covarPop - it's fine if that is 0, the slope is just 0. I think it was a typo and was meant to be only varPop.m2 checked for 0

Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/linearRegression.scala:281]: Will remove `covarPop`. 

BTW, I think this is a bug fix. Such null behaviour is consistent everywhere in other aggregate functions that use divisions, see also

```scala
scala> spark.conf.set("spark.sql.ansi.enabled", true)

scala> spark.range(1).selectExpr("sum(null)").show()
+---------+
|sum(NULL)|
+---------+
|     null|
+---------+
```

It's more null propagation that failed before. ANSI documentation doesn't mention about null handling from my cursory look so I think we can just keep ti consistent within our codebase

Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/linearRegression.scala]: At first, I think if the `covarPop.n === 0.0` means `varPop.n === 0.0` and `varPop.m2 === 0.0`.
It seems `varPop.m2 === 0.0` not determined by `varPop.n === 0.0`.

**Comment by beliefer** [comment]
LGTM. @HyukjinKwon

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**